src/compile/selection/transforms/translate.ts

```diff
@@ -0,0 +1,103 @@
+import {selector as parseSelector} from 'vega-parser';
+import {UnitModel} from './../../unit';
+import {SelectionComponent} from '../../../selection';
+import {X, Y, Channel} from '../../../channel';
+import {stringValue} from '../../../util';
+import {warn} from '../../../log';
+import {TransformCompiler} from './';
+import {default as scalesCompiler, domain} from './scales';
+import {projections as intervalProjections, NS as INTERVAL} from '../types/interval';
+
+const NS = {
+  ANCHOR: '_translate_anchor',
+  DELTA: '_translate_delta'
+};
+
+const translate:TransformCompiler = {
+  has: function(sel) {
+    return sel.translate !== undefined && sel.translate !== false;
+  },
+
+  signals: function(model, sel, signals) {
+    if (sel.type !== 'interval') {
+      warn('translate is currently only supported for interval selections.');
+      return signals;
+    }
+
+    let name = sel.name,
+        scales = scalesCompiler.has(sel),
+        size = scales ? 'unit' : name + INTERVAL.SIZE,
+        anchor = name + NS.ANCHOR,
+        events = parseSelector(sel.translate, 'scope'),
+        {x, y} = intervalProjections(sel);
+
+    if (!scales) {
+      events = events.map((e) => (e.between[0].markname = name + INTERVAL.BRUSH, e));
+    }
+
+    signals.push({
+      name: anchor,
+      value: {},
+      on: [{
+        events: events.map((e) => e.between[0]),
+        update: '{x: x(unit), y: y(unit), ' +
+          'width: ' + size + '.width, height: ' + size + '.height, ' +
+
+          (x !== null ? 'extent_x: ' + (scales ? domain(model, X) :
+              'slice(' + name + '_x)') + ', ' : '') +
+
+          (y !== null ? 'extent_y: ' + (scales ? domain(model, Y) :
+              'slice(' + name + '_y)') + ', ' : '') + '}'
+      }]
+    }, {
+      name: name + NS.DELTA,
+      value: {},
+      on: [{
+        events: events,
+        update: '{x: x(unit) - ' + anchor + '.x, y: y(unit) - ' + anchor + '.y}'
```

Comment:
This and many other string operations in this PR will be easier to read with [template strings](https://basarat.gitbooks.io/typescript/content/docs/template-strings.html)